cursive-core/src/buffer.rs

```diff
@@ -179,14 +179,32 @@ impl PrintBuffer {
 
         let mut pos = start;
 
+        fn is_control_char(g: &str) -> bool {
+            g.chars().all(|c| {
+                (
+                    c <= '\u{001F}'
+                    //&& (c != '\u{0009}' && c != '\u{000A}') //TODO: special case for tab or newline?
+                ) || (c >= '\u{007F}' && c <= '\u{009F}')
+            })
+        }
```

Comment:
I think specifically to check if it's a control char, we could be simpler and directly check the bytes: the bytes should be in the range `0x00..=0x1F | 0x7F..=0x9F`. This excludes any utf-8 continuation byte, so we already exclude any multi-byte character.

I thought initially that `g` would also be single-byte for control chars, but that's maybe not true (for example `\r\n` might be a single grapheme, and would still count as control char).

We should return `true` for newlines, tabs and such - basically anything that doesn't neatly fit in a single cell should return true here (and ideally should never be sent in the first place, but that's a topic for another day).

Reply:
Looks great otherwise! We're just waiting for ncurses itself to be updated, right?

Reply:
> Looks great otherwise! We're just waiting for ncurses itself to be updated, right?

yes, but this behavior doesn't depend on ncurses being updated at all, that is, it should actually be a separate PR and be applied on the current cursive to fix #780 else it just exits currently due to a panic:
```
    Finished dev [unoptimized + debuginfo] target(s) in 0.03s
     Running `target/x86_64-unknown-linux-gnu/debug/cursive_nul_in_text`
thread 'main' panicked at /home/user/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ncurses-5.101.0/src/lib.rs:65:29:
called `Result::unwrap()` on an `Err` value: NulError(0, [0])
stack backtrace:
   0: rust_begin_unwind
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/std/src/panicking.rs:645:5
   1: core::panicking::panic_fmt
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/panicking.rs:72:14
   2: core::result::unwrap_failed
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/result.rs:1649:5
   3: core::result::Result<T,E>::unwrap
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/result.rs:1073:23
   4: <&str as ncurses::ToCStr>::to_c_str
             at /home/user/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ncurses-5.101.0/src/lib.rs:65:9
   5: ncurses::addstr
             at /home/user/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ncurses-5.101.0/src/lib.rs:99:23
   6: <cursive::backends::curses::n::Backend as cursive_core::backend::Backend>::print
             at /tmp/cursive/cursive/src/backends/curses/n.rs:401:9
   7: cursive_core::buffer::PrintBuffer::flush
             at /tmp/cursive/cursive-core/src/buffer.rs:291:13
   8: cursive_core::cursive_run::CursiveRunner<C>::draw
             at /tmp/cursive/cursive-core/src/cursive_run.rs:98:9
   9: cursive_core::cursive_run::CursiveRunner<C>::refresh
             at /tmp/cursive/cursive-core/src/cursive_run.rs:193:9
  10: cursive_core::cursive_run::CursiveRunner<C>::run
             at /tmp/cursive/cursive-core/src/cursive_run.rs:233:9
  11: cursive_core::cursive_root::Cursive::try_run_with
             at /tmp/cursive/cursive-core/src/cursive_root.rs:898:9
  12: <cursive_core::cursive_root::Cursive as cursive::cursive_ext::CursiveExt>::run_ncurses
             at /tmp/cursive/cursive/src/cursive_ext.rs:87:9
  13: <cursive_core::cursive_root::Cursive as cursive::cursive_ext::CursiveExt>::run
             at /tmp/cursive/cursive/src/cursive_ext.rs:76:17
  14: cursive_nul_in_text::main
             at ./src/main.rs:9:5
  15: core::ops::function::FnOnce::call_once
             at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

I kind of confused things by adding this commit to this unrelated PR, I shouldn't have. If this unrelated PR(about ncurses v6) would've been applied (and ncurses been updated too) then the behavior here with `\0`(which should be in a different PR, just to be clear) would only fix an issue where the size and shadows were off:
![with_ncurses_6](https://github.com/gyscos/cursive/assets/106607441/c4790e0c-f227-4f18-a8a0-000b85181d0d)

Reply:
I cherry-picked the last commit (just removed the `\0` width test, so we're not tied to unicode-width 1.1.13)

Reply:
> because it's now only testing that the replacement char has width 1.

Which is fine: we only print that _if_ the control char has a width of 1. So:
* On old unicode-width, control chars are ignored, no replacement.
* On new unicode-width, control chars are width=1, replaced with width=1 replacement char.

To be extra robust we could check for a double-width control char (not possible today, but maybe in a future unicode-width update? :( ) and send more replacement chars (or a double-width one) in that case. Not super important though as it's not a likely scenario.

> maybe printing the control char raw like [that](https://github.com/gyscos/cursive/commit/0673fbcbed8acb78f2d1c9a11c5158b5a621e737#diff-5ab05e635edf74924d76fd94ec7bd7712e5193641e3b12b77cc47662f6d33b64R205) might not be a good idea, if it's '\r' it would jump to start of line and overwrite that part of the panic message maybe?

Ah right, making it into a debug-print instead :)

Fixed the other one too.